core/storage/file/gae_models_test.py

```diff
@@ -0,0 +1,193 @@
+            file_models.FileMetadataModel._construct_id( # pylint: disable=protected-access
+                'exp_id1', '/exp_id1/path/to/file.png'))
+
+        self.assertEqual(
+            file_metadata_model_id1, '/exp_id1/path/to/file.png')
+        self.assertEqual(
+            file_metadata_model_id2, '/exp_id1/path/to/file.png')
+
+    def test_create_creates_model_with_correct_id(self):
+        file_metadata_model = file_models.FileMetadataModel.create(
+            'exp_id1', 'path/to/file.png')
+        self.assertEqual(file_metadata_model.deleted, False)
+        self.assertEqual(
+            file_metadata_model.id, '/exp_id1/path/to/file.png')
+
+    def test_get_model_returns_the_instance_with_desired_id(self):
+        file_metadata_model = file_models.FileMetadataModel.create(
+            'exp_id1', 'path/to/file1.png')
+        file_metadata_model.commit(feconf.SYSTEM_COMMITTER_ID, [])
+        actual_model = file_models.FileMetadataModel.get_model(
+            'exp_id1', 'path/to/file1.png')
+
+        self.assertEqual(actual_model, file_metadata_model)
+
+        actual_model = file_models.FileMetadataModel.get_model(
+            'exp_id1', 'path/to/file2.png')
+        self.assertEqual(actual_model, None)
+
+        with self.assertRaisesRegexp(Exception, (
+            'Entity for class FileMetadataModel '
+            'with id /exp_id1/path/to/file2.png not found')):
+            actual_model = file_models.FileMetadataModel.get_model(
+                'exp_id1', 'path/to/file2.png', True)
+
+    def test_get_version_returns_instance_with_desired_version(self):
+        file_metadata_model = file_models.FileMetadataModel.create(
+            'exp_id1', 'path/to/file1.png')
+        file_metadata_model.commit(feconf.SYSTEM_COMMITTER_ID, [])
+
+        actual_model = file_models.FileMetadataModel.get_version(
+            'exp_id1', 'path/to/file1.png', 1)
+        self.assertEqual(file_metadata_model.key, actual_model.key)
+
+        file_metadata_model.commit(feconf.SYSTEM_COMMITTER_ID, [])
+        actual_model = file_models.FileMetadataModel.get_version(
+            'exp_id1', 'path/to/file1.png', 2)
+        self.assertEqual(file_metadata_model.key, actual_model.key)
+
+    def test_commit_updates_version_of_instance(self):
+        file_metadata_model = file_models.FileMetadataModel.create(
+            'exp_id1', 'path/to/file1.png')
+        self.assertEqual(file_metadata_model.version, 0)
+
+        file_metadata_model.commit(feconf.SYSTEM_COMMITTER_ID, [])
+        self.assertEqual(file_metadata_model.version, 1)
+
+
+class FileModelTest(test_utils.GenericTestBase):
+    """Tests the FileModel class."""
+
+    def test_reconstitute_changes_the_content_snapshot(self):
+        file_model = file_models.FileModel.create(
+            'exp_id1', 'path/to/file.png')
+        self.assertEqual(file_model.content, None)
+        file_model._reconstitute('file_content') # pylint: disable=protected-access
```

Comment:
Ditto here and elsewhere--prefer not testing the private API of a class as it can lead to brittle tests that fail in unexpected situations (such as when replacing the implementation of an API with something new but keeping the interface and behaviors the same).

Reply:
Hmm, since the file model overrides _compute_snapshot, the resulting effect is that file model snapshots include the created_on and last_updated properties. Could we maybe reconstitute an old snapshot of a particular file model and verify that it includes those properties? Another thing we might be able to do: use blame to figure out why the method was originally overloaded, and then see if whatever the reason is can be turned into a specific behavior to test.

One way to verify that whatever test you come up with works is by removing the overloaded method to ensure the new test(s) fail(s) as expected.

Reply:
Hmm, since the file model overrides _compute_snapshot, the resulting effect is that file model snapshots include the created_on and last_updated properties. Could we maybe reconstitute an old snapshot of a particular file model and verify that it includes those properties? Another thing we might be able to do: use blame to figure out why the method was originally overloaded, and then see if whatever the reason is can be turned into a specific behavior to test.

One way to verify that whatever test you come up with works is by removing the overloaded method to ensure the new test(s) fail(s) as expected.

Reply:
The superclass method returns a dict with keys `content`, `deleted` and `version`. The overloaded method returns the content of the file model which is of type str. 

The function is used when a commit is made to update the model contents (https://github.com/oppia/oppia/blob/develop/core/storage/base_model/gae_models.py#L524)

I added a test which updates the file contents in  a commit and checks if the content is updated correctly.

Also, after removing the overloaded function, a BadValueError is raised since content is expected to be of type str but a dict is received.

Does the testing approach look fine?